### PR TITLE
Fix broken curl transport on setting up url and broken retry policy deleting all headers

### PR DESCRIPTION
**Pull request by vhvb1989**
Two important fixes here
- Curl transport was appending a '0' and not a \0 at the end of url, this was making curl to write beyond url and returning not found as response all the time

- retry policy was removing all current headers as if they all were retry headers. Fixed by calling set_retry_start as first thing for retry policy

### CODE REVIEW

File: sdk/core/core/src/az_aad.c [hunk 1]
```diff
@@ -100,7 +100,7 @@ AZ_NODISCARD az_result _az_aad_request_token(_az_http_request* request, _az_toke
   AZ_RETURN_IF_FAILED(az_http_request_append_header(
       request,
       AZ_SPAN_FROM_STR("Content-Type"),
-      AZ_SPAN_FROM_STR("application/x-www-url-form-urlencoded")));
```
Comment (antkmsft) [inline sdk/core/core/src/az_aad.c:103]: Ha, this is where their docs and postman sample they provide disagree (https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow). But I'd go with your change, assuming that you verified that it works.

File: sdk/core/core/src/az_http_policy_retry.c [hunk 2]
```diff
@@ -107,7 +107,7 @@ AZ_INLINE AZ_NODISCARD az_result _az_http_policy_retry_get_retry_after(
     {
       if (az_span_is_content_equal_ignoring_case(header.key, AZ_SPAN_FROM_STR("retry-after-ms"))
           || az_span_is_content_equal_ignoring_case(
-              header.key, AZ_SPAN_FROM_STR("x-ms-retry-after-ms")))
+                 header.key, AZ_SPAN_FROM_STR("x-ms-retry-after-ms")))
```
Comment (antkmsft) [inline sdk/core/core/src/az_http_policy_retry.c:110]: is this space coming from Ctrl+K+D (auto-format)?

File: sdk/core/core/src/az_aad.c (hunk 1 above)
Comment (vhvb1989) [reply sdk/core/core/src/az_aad.c:103]: yeah, we got an error doing it the other way.
I see their doc with this value as well: https://github.com/Azure/azure-sdk-for-c/pull/573

So, it was just a typo somewhere

File: sdk/core/core/src/az_http_policy_retry.c (hunk 2 above)
Comment (vhvb1989) [reply sdk/core/core/src/az_http_policy_retry.c:110]: yes Sr.

File: sdk/platform/http_client/curl/src/az_curl.c [hunk 3]
```diff
@@ -109,7 +109,7 @@ _az_span_append_header_to_buffer(az_span writable_buffer, az_pair header, az_spa
   writable_buffer = az_span_append(writable_buffer, header.key);
   writable_buffer = az_span_append(writable_buffer, separator);
   writable_buffer = az_span_append(writable_buffer, header.value);
-  writable_buffer = az_span_append_uint8(writable_buffer, '0');
+  writable_buffer = az_span_append_uint8(writable_buffer, 0);
```
Comment (ahsonkhan) [inline sdk/platform/http_client/curl/src/az_curl.c:112]: Looks good. Nice catch!

**Pull request merged**